torch/distributed/_sharded_tensor/__init__.py

```diff
@@ -510,3 +511,109 @@ def shard_parameter(
+    parallelism.
+
+    Args:
+        sharded_tensor: Sharded tensor which the op was performed on.
+        local_result: A tensor which is from the op performed on the local_shard of
+            the sharded_tensor.
+        tensor_shard_dim: Dim which the tensor is sharded on.
+        result_shard_dim: Dim which the new sharded tensor will be sharded on.
+        world_size: number of ranks.
+        pg (ProcessGroup, optional): The process group to work on. If None,
+            the default process group will be used.
+
+    Return: new sharded tensor from the local_result.
+    """
+    sharded_weight_metadata = sharded_tensor.local_shards()[0].metadata
+    current_offsets = [0] * len(local_result.size())
+    current_offsets[result_shard_dim] = sharded_weight_metadata.shard_offsets[tensor_shard_dim]
+    local_shard_metadata = ShardMetadata(
+        shard_offsets=current_offsets,
+        shard_sizes=list(local_result.size()),
+        placement=sharded_weight_metadata.placement,
+    )
+    local_shards = [Shard(local_result, local_shard_metadata)]
+    global_size = list(local_result.size())
+    global_size[result_shard_dim] = sharded_tensor.size(tensor_shard_dim)
+
+    new_st = ShardedTensor._init_from_local_shards(local_shards, tuple(global_size), process_group=pg)
+
+    # Manually set sharding_spec
+    new_st._sharding_spec = copy.deepcopy(sharded_tensor._sharding_spec)
+    new_st._sharding_spec.dim = result_shard_dim
+    return new_st
+
+
+def merge_sharded_local_results(
+    sharded_local_result,
+    world_size,
+    pg=None,
+):
+    """
+    Given a local result sharded tensor, we want to merge it into a local tensor.
+
+    Args:
+        sharded_local_result: Sharded tensor which we need to merge.
+        world_size: number of ranks.
+        pg (ProcessGroup, optional): The process group to work on. If None,
+            the default process group will be used.
+
+    Return: new sharded tensor from the local_result.
+    """
+    from torch.distributed._sharded_tensor.ops._common import (
+        _result_distribute_with_col_rearrange,
+    )
+    local_shard = sharded_local_result.local_shards()[0].tensor
+    local_results = []
+    for i in range(local_shard.size(0)):
+        local_results.append(local_shard[i].t().contiguous())
+
+    # Distribute results to each rank with col rearrangement.
+    output = _result_distribute_with_col_rearrange(
+        local_results,
+        local_shard,
+        world_size,
+        sharded_local_result,
+        pg
+    )
+
+    # transpose the output and return result.
+    return output.t().contiguous()
+
+
+def aggregate_partial_tensor_list(
+    tensor_list,
+    rank,
+    pg=None,
+):
+    """
+    Given parital result from op on a sharded tensor. We want to aggregate the
+    partial result to create a fully synced local tensor. This can seen as the
+    last step of the second phase of the Megatron-LM style model(tensor) parallelism.
```

Comment:
This is ok for a prototype, but again I think for a proper API we should have a `PartialTensor` class like `ShardedTensor` and have ops on it to reshard and aggregate appropriately.

I'd suggest writing a design doc for `PartialTensor` based on the use case we have here and then implementing it.